Upload blog post edit (#19)
CorneliaKelinske/first_blog: posts/2022/10-01-api_queries_with_finch.md

==title==
API queries with Finch

==author==
Cornelia Kelinske

==description==
I've been using HTTPoison as my HTTP client of choice ever since I learned how to do API queries with Elixir.
Lately, though, I've been hearing more about people using Finch for that purpose. So I went ahead and tried it out.

==tags==
coding, elixir, api_calls, finch

==body==

# 1. The set-up

> The first thing that is good to know about Finch is that we want to add it to our supervision tree. This means that, ideally, we'll remember to create our new app with the `mix new new_app --sup` command so that we get our `application.ex` file with the supervision tree automatically generated.

Reviewer: ```suggestion
The first thing that is good to know about Finch is that we want to add it to our supervision tree. Generate a new project with `mix new new_app --sup` command so that we get our `application.ex` file with the supervision tree automatically generated.
```

Then, as with all libraries, we need to head over to the [Finch documentation](https://hexdocs.pm/finch/Finch.html).

With the documentation open, we can then add Finch to our dependencies in the `mix.exs` file

> ```

Reviewer: You can add elixir to the end of these 3 backticks

Of course, we could just extend the `query_api/1` function above and pass in all the other params. 
But that would make for some really ugly code (at least in my opinion). So let's stay away from that.

Let's take a more dynamic approach instead!

I'm currently big into writing helper modules, so let's put our query builder functions into a helper module.

```
defmodule NewApp.FinchHelpers do
  @moduledoc """
  Functions for building an API call through Finch
  """
  @type params :: %{api_key: String.t() | nil, q: String.t() | nil, limit: pos_integer} | []
  

  @spec build_query(params(), String.t()) :: Finch.Request.t()
  def build_query(params, url) do
    url
    |> append_params(params)
    |> String.trim_trailing("&")
    |> then(fn x -> Finch.build(:get, x) end)
  end

  def append_params(query, []) do
    query
  end

  def append_params(query, params) do
    Enum.reduce(params, "#{query}?", fn {k, v}, acc -> acc <> "#{k}=#{v}&" end)
  end
```

> In this example, we are adding an API key, a query term (q) and a limit to our URL. All the params are passed into our main `build_query/2` function in a map so that both keys and values can be used for building the query.

Reviewer: ```suggestion
In this example, we are adding an API key, the query variable `q` and a limit to our URL. All the parameters are passed into our main `build_query/2` function in a map so that both keys and values can be used for building the query.
```
